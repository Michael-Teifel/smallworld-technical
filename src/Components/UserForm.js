import { useEffect, useState } from 'react';
import { useForm } from 'react-hook-form';
import { queryAPI } from './../API/form'
import AlertMessage from './AlertMessage';

function UserForm() {
  // Declare Component States
  const [formData, setFormData] = useState({
    jobFunctions: [],
    states: []
  });
  const [showAlertMessage, setShowAlertMessage] = useState(false);
  const [alertMessageProps, setAlertMessageProps] = useState({});
  
  // Register the useForm hook.
  const { 
    register, 
    handleSubmit,
    reset,
    formState: {isSubmitSuccessful} } = useForm();
  

  // Initialize the default form values
  useEffect(() => {
    let response = queryAPI("https://not-a-real-endpoint.smallworld.ai/form", "GET")
    response = JSON.parse(response);
    if(response.status === 200) {
      setFormData(response.data);
    } 
  }, []);

  // Reset form to the default values if it is submitted succesfully
  useEffect(() => {
    if (isSubmitSuccessful) {
      reset();
    }
  }, [isSubmitSuccessful, reset]);
  
  // Form Submission
  function onSubmit(data) {
    let response = queryAPI("https://not-a-real-endpoint.smallworld.ai/form", "POST", data)
    response = JSON.parse(response);
    if(response.status === 200) {

      setShowAlertMessage(true);
      setAlertMessageProps({
        messageType: 'success',
        message: 'Form was succesfully submitted'
      })
      
    } else {
      setShowAlertMessage(true);
      setAlertMessageProps({
        messageType: 'error',
        message: 'An unkown server error occurred'
      })
    }
  }
  
  // JSX
  const jobFunctionElements = formData.jobFunctions.map(jobFunction => {
    return <option key={jobFunction} value={jobFunction}>{jobFunction}</option>;
  });
  
  const stateElements = formData.states.map(state => {
    return <option key={state.abbreviation} value={state.name}>{state.abbreviation}</option>;
  });

  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      {showAlertMessage ? <AlertMessage {...alertMessageProps} /> : undefined}
      <label>
        First Name:
        <input type="text" {...register("firstName", {required: true})} />
      </label>
      <label>
        Last Name:
        <input type="text" {...register("lastName", {required: true})} />
      </label>
      <label>
        Email Address:
        <input type="email"{...register("emailAddress", {required: true})} />
      </label>
      <label>
        Password:
        <input type="password" {...register("password", {required: true})} />
      </label>
      <label>
        Company:
        <input type="text" {...register("company", {required: true})} />
      </label>
      <label>
        Job Title:
        <input type="text" {...register("jobTitle", {required: true})} />
      </label>
      <label>
        Job Function:
        <select defaultValue={""}{...register("jobFunction", {required: true})}>
          <option value="" disabled>Select an Option</option>
          {jobFunctionElements}
        </select>
      </label>
      <label>
        State:
        <select defaultValue={""} {...register("state", {required: true})}>
          <option value="" disabled>Select an Option</option>
          {stateElements}
        </select>
      </label>
      <label>
        City:
        <input type="text" {...register("city", {required: true})} />
      </label>
      
      <input type="submit" value="Submit" />
    </form>
  );
}

export default UserForm;