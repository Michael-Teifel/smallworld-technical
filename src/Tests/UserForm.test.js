import { fireEvent, render, screen } from '@testing-library/react';
import UserForm from '../Components/UserForm';


it('renders the UserForm', () => {
  const {asFragment} = render(<UserForm />);
  expect(asFragment()).toMatchSnapshot();
});

it('submits the form if fields are complete', () => {
  const {asFragment} = render(<UserForm />);

  const firstName = screen.getByLabelText('First Name:');
  fireEvent.change(firstName, {target: {value: 'Alex'}});

  const lastName = screen.getByLabelText('Last Name:');
  fireEvent.change(lastName, {target: {value: 'Honnold'}});

  const email = screen.getByLabelText('Email Address:');
  fireEvent.change(email, {target: {value: 'Alex@sbc.com'}});

  const password = screen.getByLabelText('Password:');
  fireEvent.change(password, {target: {value: 'js3Qr'}});

  const company = screen.getByLabelText('Company:');
  fireEvent.change(company, {target: {value: 'Planet Granite'}});

  const jobTitle = screen.getByLabelText('Job Title:');
  fireEvent.change(jobTitle, {target: {value: 'Climbing Instructor'}});

  const jobFunction = screen.getByLabelText('Job Function:');
  fireEvent.change(jobFunction, {target: {value: 'Sales'}});

  const state = screen.getByLabelText('State:');
  fireEvent.change(state, {target: {value: 'Nevada'}});

  const city = screen.getByLabelText('City:');
  fireEvent.change(city, {target: {value: 'Yosemite'}});

  const submit = screen.getByText('Submit');
  fireEvent.click(submit)


  expect(asFragment()).toMatchSnapshot();
});

it('does not submit the form if jobFunction and State Are left as default', () => {
  const {asFragment} = render(<UserForm />);

  const firstName = screen.getByLabelText('First Name:');
  fireEvent.change(firstName, {target: {value: 'Alex'}});

  const lastName = screen.getByLabelText('Last Name:');
  fireEvent.change(lastName, {target: {value: 'Honnold'}});

  const email = screen.getByLabelText('Email Address:');
  fireEvent.change(email, {target: {value: 'Alex@sbc.com'}});

  const password = screen.getByLabelText('Password:');
  fireEvent.change(password, {target: {value: 'js3Qr'}});

  const company = screen.getByLabelText('Company:');
  fireEvent.change(company, {target: {value: 'Planet Granite'}});

  const jobTitle = screen.getByLabelText('Job Title:');
  fireEvent.change(jobTitle, {target: {value: 'Climbing Instructor'}});

  const city = screen.getByLabelText('City:');
  fireEvent.change(city, {target: {value: 'Yosemite'}});

  const submit = screen.getByText('Submit');
  fireEvent.click(submit)


  expect(asFragment()).toMatchSnapshot();
});

it('does not submit if password is missing', () => {
  const {asFragment} = render(<UserForm />);

  const firstName = screen.getByLabelText('First Name:');
  fireEvent.change(firstName, {target: {value: 'Alex'}});

  const lastName = screen.getByLabelText('Last Name:');
  fireEvent.change(lastName, {target: {value: 'Honnold'}});

  const email = screen.getByLabelText('Email Address:');
  fireEvent.change(email, {target: {value: 'Alex@sbc.com'}});

  const company = screen.getByLabelText('Company:');
  fireEvent.change(company, {target: {value: 'Planet Granite'}});

  const jobTitle = screen.getByLabelText('Job Title:');
  fireEvent.change(jobTitle, {target: {value: 'Climbing Instructor'}});

  const jobFunction = screen.getByLabelText('Job Function:');
  fireEvent.change(jobFunction, {target: {value: 'Sales'}});

  const state = screen.getByLabelText('State:');
  fireEvent.change(state, {target: {value: 'Nevada'}});

  const city = screen.getByLabelText('City:');
  fireEvent.change(city, {target: {value: 'Yosemite'}});

  const submit = screen.getByText('Submit');
  fireEvent.click(submit)


  expect(asFragment()).toMatchSnapshot();
});

it('does not submit if email is invalid', () => {
  const {asFragment} = render(<UserForm />);

  const firstName = screen.getByLabelText('First Name:');
  fireEvent.change(firstName, {target: {value: 'Alex'}});

  const lastName = screen.getByLabelText('Last Name:');
  fireEvent.change(lastName, {target: {value: 'Honnold'}});

  const email = screen.getByLabelText('Email Address:');
  fireEvent.change(email, {target: {value: 'Alex'}});

  const password = screen.getByLabelText('Password:');
  fireEvent.change(password, {target: {value: 'js3Qr'}});

  const company = screen.getByLabelText('Company:');
  fireEvent.change(company, {target: {value: 'Planet Granite'}});

  const jobTitle = screen.getByLabelText('Job Title:');
  fireEvent.change(jobTitle, {target: {value: 'Climbing Instructor'}});

  const jobFunction = screen.getByLabelText('Job Function:');
  fireEvent.change(jobFunction, {target: {value: 'Sales'}});

  const state = screen.getByLabelText('State:');
  fireEvent.change(state, {target: {value: 'Nevada'}});

  const city = screen.getByLabelText('City:');
  fireEvent.change(city, {target: {value: 'Yosemite'}});

  const submit = screen.getByText('Submit');
  fireEvent.click(submit)


  expect(asFragment()).toMatchSnapshot();
});