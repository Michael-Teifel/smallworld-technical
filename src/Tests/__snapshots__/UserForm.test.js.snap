// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`does not submit if email is invalid 1`] = `
<DocumentFragment>
  <form>
    <label>
      First Name:
      <input
        name="firstName"
        type="text"
      />
    </label>
    <label>
      Last Name:
      <input
        name="lastName"
        type="text"
      />
    </label>
    <label>
      Email Address:
      <input
        name="emailAddress"
        type="email"
      />
    </label>
    <label>
      Password:
      <input
        name="password"
        type="password"
      />
    </label>
    <label>
      Company:
      <input
        name="company"
        type="text"
      />
    </label>
    <label>
      Job Title:
      <input
        name="jobTitle"
        type="text"
      />
    </label>
    <label>
      Job Function:
      <select
        name="jobFunction"
      >
        <option
          disabled=""
          selected=""
          value=""
        >
          Select an Option
        </option>
        <option
          value="Accounting"
        >
          Accounting
        </option>
        <option
          value="Sales"
        >
          Sales
        </option>
        <option
          value="Marketing"
        >
          Marketing
        </option>
      </select>
    </label>
    <label>
      State:
      <select
        name="state"
      >
        <option
          disabled=""
          selected=""
          value=""
        >
          Select an Option
        </option>
        <option
          value="California"
        >
          CA
        </option>
        <option
          value="Nevada"
        >
          NV
        </option>
      </select>
    </label>
    <label>
      City:
      <input
        name="city"
        type="text"
      />
    </label>
    <input
      type="submit"
      value="Submit"
    />
  </form>
</DocumentFragment>
`;

exports[`does not submit if password is missing 1`] = `
<DocumentFragment>
  <form>
    <label>
      First Name:
      <input
        name="firstName"
        type="text"
      />
    </label>
    <label>
      Last Name:
      <input
        name="lastName"
        type="text"
      />
    </label>
    <label>
      Email Address:
      <input
        name="emailAddress"
        type="email"
      />
    </label>
    <label>
      Password:
      <input
        name="password"
        type="password"
      />
    </label>
    <label>
      Company:
      <input
        name="company"
        type="text"
      />
    </label>
    <label>
      Job Title:
      <input
        name="jobTitle"
        type="text"
      />
    </label>
    <label>
      Job Function:
      <select
        name="jobFunction"
      >
        <option
          disabled=""
          selected=""
          value=""
        >
          Select an Option
        </option>
        <option
          value="Accounting"
        >
          Accounting
        </option>
        <option
          value="Sales"
        >
          Sales
        </option>
        <option
          value="Marketing"
        >
          Marketing
        </option>
      </select>
    </label>
    <label>
      State:
      <select
        name="state"
      >
        <option
          disabled=""
          selected=""
          value=""
        >
          Select an Option
        </option>
        <option
          value="California"
        >
          CA
        </option>
        <option
          value="Nevada"
        >
          NV
        </option>
      </select>
    </label>
    <label>
      City:
      <input
        name="city"
        type="text"
      />
    </label>
    <input
      type="submit"
      value="Submit"
    />
  </form>
</DocumentFragment>
`;

exports[`does not submit the form if jobFunction and State Are left as default 1`] = `
<DocumentFragment>
  <form>
    <label>
      First Name:
      <input
        name="firstName"
        type="text"
      />
    </label>
    <label>
      Last Name:
      <input
        name="lastName"
        type="text"
      />
    </label>
    <label>
      Email Address:
      <input
        name="emailAddress"
        type="email"
      />
    </label>
    <label>
      Password:
      <input
        name="password"
        type="password"
      />
    </label>
    <label>
      Company:
      <input
        name="company"
        type="text"
      />
    </label>
    <label>
      Job Title:
      <input
        name="jobTitle"
        type="text"
      />
    </label>
    <label>
      Job Function:
      <select
        name="jobFunction"
      >
        <option
          disabled=""
          selected=""
          value=""
        >
          Select an Option
        </option>
        <option
          value="Accounting"
        >
          Accounting
        </option>
        <option
          value="Sales"
        >
          Sales
        </option>
        <option
          value="Marketing"
        >
          Marketing
        </option>
      </select>
    </label>
    <label>
      State:
      <select
        name="state"
      >
        <option
          disabled=""
          selected=""
          value=""
        >
          Select an Option
        </option>
        <option
          value="California"
        >
          CA
        </option>
        <option
          value="Nevada"
        >
          NV
        </option>
      </select>
    </label>
    <label>
      City:
      <input
        name="city"
        type="text"
      />
    </label>
    <input
      type="submit"
      value="Submit"
    />
  </form>
</DocumentFragment>
`;

exports[`renders the UserForm 1`] = `
<DocumentFragment>
  <form>
    <label>
      First Name:
      <input
        name="firstName"
        type="text"
      />
    </label>
    <label>
      Last Name:
      <input
        name="lastName"
        type="text"
      />
    </label>
    <label>
      Email Address:
      <input
        name="emailAddress"
        type="email"
      />
    </label>
    <label>
      Password:
      <input
        name="password"
        type="password"
      />
    </label>
    <label>
      Company:
      <input
        name="company"
        type="text"
      />
    </label>
    <label>
      Job Title:
      <input
        name="jobTitle"
        type="text"
      />
    </label>
    <label>
      Job Function:
      <select
        name="jobFunction"
      >
        <option
          disabled=""
          selected=""
          value=""
        >
          Select an Option
        </option>
        <option
          value="Accounting"
        >
          Accounting
        </option>
        <option
          value="Sales"
        >
          Sales
        </option>
        <option
          value="Marketing"
        >
          Marketing
        </option>
      </select>
    </label>
    <label>
      State:
      <select
        name="state"
      >
        <option
          disabled=""
          selected=""
          value=""
        >
          Select an Option
        </option>
        <option
          value="California"
        >
          CA
        </option>
        <option
          value="Nevada"
        >
          NV
        </option>
      </select>
    </label>
    <label>
      City:
      <input
        name="city"
        type="text"
      />
    </label>
    <input
      type="submit"
      value="Submit"
    />
  </form>
</DocumentFragment>
`;

exports[`submits the form if fields are complete 1`] = `
<DocumentFragment>
  <form>
    <label>
      First Name:
      <input
        name="firstName"
        type="text"
      />
    </label>
    <label>
      Last Name:
      <input
        name="lastName"
        type="text"
      />
    </label>
    <label>
      Email Address:
      <input
        name="emailAddress"
        type="email"
      />
    </label>
    <label>
      Password:
      <input
        name="password"
        type="password"
      />
    </label>
    <label>
      Company:
      <input
        name="company"
        type="text"
      />
    </label>
    <label>
      Job Title:
      <input
        name="jobTitle"
        type="text"
      />
    </label>
    <label>
      Job Function:
      <select
        name="jobFunction"
      >
        <option
          disabled=""
          selected=""
          value=""
        >
          Select an Option
        </option>
        <option
          value="Accounting"
        >
          Accounting
        </option>
        <option
          value="Sales"
        >
          Sales
        </option>
        <option
          value="Marketing"
        >
          Marketing
        </option>
      </select>
    </label>
    <label>
      State:
      <select
        name="state"
      >
        <option
          disabled=""
          selected=""
          value=""
        >
          Select an Option
        </option>
        <option
          value="California"
        >
          CA
        </option>
        <option
          value="Nevada"
        >
          NV
        </option>
      </select>
    </label>
    <label>
      City:
      <input
        name="city"
        type="text"
      />
    </label>
    <input
      type="submit"
      value="Submit"
    />
  </form>
</DocumentFragment>
`;
